hip/solver/common_trs_kernels.hip.hpp

```diff
@@ -0,0 +1,244 @@
+
+
+#include <functional>
+#include <memory>
+
+
+#include <hip/hip_runtime.h>
+#include <hipsparse.h>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+
+
+#include "core/matrix/dense_kernels.hpp"
+#include "core/synthesizer/implementation_selection.hpp"
+#include "hip/base/device_guard.hip.hpp"
+#include "hip/base/hipsparse_bindings.hip.hpp"
+#include "hip/base/math.hip.hpp"
+#include "hip/base/pointer_mode_guard.hip.hpp"
+#include "hip/base/types.hip.hpp"
+
+
+namespace gko {
+namespace solver {
+
+
+struct SolveStruct {};
+
+
+namespace hip {
+
+
+struct SolveStruct : gko::solver::SolveStruct {
+    csrsv2Info_t solve_info;
+    hipsparseSolvePolicy_t policy;
+    hipsparseMatDescr_t factor_descr;
+    int factor_work_size;
+    void *factor_work_vec;
+    SolveStruct()
+    {
+        factor_work_vec = nullptr;
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(hipsparseCreateMatDescr(&factor_descr));
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(
+            hipsparseSetMatIndexBase(factor_descr, HIPSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(
+            hipsparseSetMatType(factor_descr, HIPSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(hipsparseSetMatDiagType(
+            factor_descr, HIPSPARSE_DIAG_TYPE_NON_UNIT));
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(hipsparseCreateCsrsv2Info(&solve_info));
+        policy = HIPSPARSE_SOLVE_POLICY_USE_LEVEL;
+    }
+
+    SolveStruct(const SolveStruct &) : SolveStruct() {}
+
+    SolveStruct(SolveStruct &&) : SolveStruct() {}
+
+    SolveStruct &operator=(const SolveStruct &) { return *this; }
+
+    SolveStruct &operator=(SolveStruct &&) { return *this; }
+
+    ~SolveStruct()
+    {
+        hipsparseDestroyMatDescr(factor_descr);
+        if (solve_info) {
+            hipsparseDestroyCsrsv2Info(solve_info);
+        }
+        if (factor_work_vec != nullptr) {
+            hipFree(factor_work_vec);
+            factor_work_vec = nullptr;
+        }
+    }
+};
+
+
+}  // namespace hip
+}  // namespace solver
+
+
+namespace kernels {
+namespace hip {
+namespace {
+
+
+void should_perform_transpose_kernel(std::shared_ptr<const HipExecutor> exec,
+                                     bool &do_transpose)
+{
+    do_transpose = true;
+}
+
+
+void init_struct_kernel(std::shared_ptr<const HipExecutor> exec,
+                        std::shared_ptr<solver::SolveStruct> &solve_struct)
+{
+    solve_struct = std::dynamic_pointer_cast<solver::SolveStruct>(
+        std::make_shared<solver::hip::SolveStruct>());
+}
+
+
+template <typename ValueType, typename IndexType>
+void generate_kernel(std::shared_ptr<const HipExecutor> exec,
+                     const matrix::Csr<ValueType, IndexType> *matrix,
+                     solver::SolveStruct *solve_struct,
+                     const gko::size_type num_rhs, bool is_upper)
+{
+    auto hip_solve_struct =
+        reinterpret_cast<solver::hip::SolveStruct *>(solve_struct);
```

Comment:
Shouldn't this be a `dynamic_cast`? Down-casting using `reinterpret_cast` could be very dangerous, as it ignores the data layout completely.